### PR TITLE
added mentee to apply for mentorship

### PR DESCRIPTION
**Pull request by neymarjimoh**
- mentees can apply for mentorship and choose a track
- requires track/stack to be available
- added a PR template for proper PR documentation

## How to test
- Clone, install packages
- checkout this branch
- start dev server , `npm run dev`
- test on any restful client

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,7 +27,9 @@
     "bcryptjs": "^2.4.3",
     "body-parser": "^1.19.0",
     "chalk": "^3.0.0",
+    "dotenv": "^8.2.0",
     "express": "^4.17.1",
+    "express-validator": "^6.7.0",
```
Comment (Nicanor008) [inline package.json:32]: Why we using two validators express-validator and validator?

File: src/controllers/mentee.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        })
+
+        return responseHandler(
+            res,
+            201,
+            application,
+            'Application sent. Check email for more.'
+        )
+    } catch (error) {
+        next(new InternalServerError(error))
```
Comment (Nicanor008) [inline src/controllers/mentee.js:62]: Good error catching

File: src/middlewares/validation.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { validationResult } from 'express-validator'
+import { CustomError } from '../utils/customError'
+
+export const validationMW = (req, res, next) => {
```
Comment (Nicanor008) [inline src/middlewares/validation.js:4]: what does MW stands for, any better way to write this functionName?

File: src/routes/mentees/auth.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * mentorship application for mentee
+ */
+router.post(
+    '/apply',
+    validations.menteeApplyRules(),
+    validationMW,
+    applyForMentorship
+)
+
+router.get('/ddd', (req, res) => {
```
Comment (Nicanor008) [inline src/routes/mentees/auth.js:18]: naming?? ddd???

File: src/utils/customError.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+export class CustomError extends Error {
+    constructor(statusCode, message, errors) {
```
Comment (Nicanor008) [inline src/utils/customError.js:2]: Awesome

**Review by Nicanor008 — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (neymarjimoh) [reply package.json:32]: I used it cos express-validator and joi worked well and are easy to implement in express applications

File: src/middlewares/validation.js (hunk 3 above)
Comment (neymarjimoh) [reply src/middlewares/validation.js:4]: Ok.. would change it..
Couldn't think of a better name, it stands for middle ware to return validation error responses or next chain of request

File: src/routes/mentees/auth.js (hunk 4 above)
Comment (neymarjimoh) [reply src/routes/mentees/auth.js:18]: Totally forgot to remove this.. it was the first endpoint I created to test if the mentees routes were working

File: src/utils/customError.js (hunk 5 above)
Comment (neymarjimoh) [reply src/utils/customError.js:2]: Thanks

File: src/index.js [hunk 6]
```diff
@@ -48,8 +51,21 @@ mongoose.connection.on('error', (err) => {
 
 app.use('/api/v1/users', users)
 app.use('/verify-email', emailVerify)
+app.use(mentees);
```
Comment (Nicanor008) [inline src/index.js:54]: This is different from the other(line 52, 53). Could we stick to one style then maybe refactor later.

File: src/index.js [hunk 7]
```diff
@@ -4,7 +4,10 @@ const morgan = require('morgan')
 const chalk = require('chalk')
 const mongoose = require('mongoose')
 const users = require('./routes/api/users')
+import errorHandler from './utils/errorHandler'
+
 import emailVerify from './routes/endpoints/users'
+import mentees from "./routes/mentees";
```
Comment (Nicanor008) [inline src/index.js:10]: Move the mentees folder inside the endpoints/api.
```suggestion
import mentees from "./routes/api/mentees";
```

File: src/models/track.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        ],
+        mentors: [
+            {
+                type: Schema.Types.ObjectId,
+                ref: 'User',
+            },
+        ],
+        applications: [
+            {
+                type: Schema.Types.ObjectId,
```
Comment (Nicanor008) [inline src/models/track.js:25]: I don't understand `track.js`. Is it list learning tracks available?

**Review by Nicanor008 — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 6 above)
Comment (neymarjimoh) [reply src/index.js:54]: Okay n

**Comment by Nicanor008** [comment]
Kindly adopt the new folder structure, and you can also use the /tracks endpoints to retrieve all tracks and add to the application_schema. Note that, this feature is required by tomorrow, Tuesday.

**Comment by neymarjimoh** [comment]
> Kindly adopt the new folder structure, and you can also use the /tracks endpoints to retrieve all tracks and add to the application_schema. Note that, this feature is required by tomorrow, Tuesday.

alright would do so

**Review by Nicanor008 — APPROVED** [review]

**Pull request merged**